src/utilities/petscSupport.cpp

```diff
@@ -0,0 +1,1093 @@
+                PetscReal cellCenter[dim];
+                PetscReal cellEdge[dim];  // Vector from cell center to edge center
+                PetscReal cellFace[dim];  // Vector from cell center to face center
+                PetscReal n[dim];         // Surface area normal. Needs to be corrected to align with the vector from the vertex to the edgeCenter
+
+                PetscCall(DMPlexComputeCellGeometryFVM(dm, cells[c], NULL, cellCenter, NULL));
+
+                for (PetscInt d = 0; d < dim; ++d) {
+                    cellEdge[d] = edgeCenter[d] - cellCenter[d];
+                    cellFace[d] = faceCenter[d] - cellCenter[d];
+                }
+
+                // Surface area normal. Take into account the 0.5 needed to compute the area of the triangle
+                n[0] = 0.5 * (cellEdge[1] * cellFace[2] - cellEdge[2] * cellFace[1]);
+                n[1] = 0.5 * (cellEdge[2] * cellFace[0] - cellEdge[0] * cellFace[2]);
+                n[2] = 0.5 * (cellEdge[0] * cellFace[1] - cellEdge[1] * cellFace[0]);
+
+                // The sign of the dot product of surface area normal and vector connecting vertex to edge center gives the correction
+                PetscReal sgn = PetscSignReal(vertexEdge[0] * n[0] + vertexEdge[1] * n[1] + vertexEdge[2] * n[2]);
+
+                for (PetscInt d = 0; d < dim; ++d) {
+                    N[d] += sgn * n[d];
+                }
+            }
+        }
+    }
+
+    PetscFunctionReturn(PETSC_SUCCESS);
+}
+
+// Compute the edge surface area normal as defined in Morgan and Waltz with respect to a given vertex and an edge center
+// NOTE: This does NOT check if the vertex and edge are actually associated with each other.
+static PetscErrorCode DMPlexEdgeSurfaceAreaNormal(DM dm, const PetscInt v, const PetscInt e, PetscReal N[]) {
+    PetscFunctionBegin;
+
+    PetscReal edgeCenter[3], vCoords[3];
+    PetscInt dim;
+    PetscInt nFace;
+    const PetscInt *faces;
+
+    PetscCall(DMGetDimension(dm, &dim));
+
+    PetscCall(DMPlexComputeCellGeometryFVM(dm, v, NULL, vCoords, NULL));
+    PetscCall(DMPlexComputeCellGeometryFVM(dm, e, NULL, edgeCenter, NULL));
+
+    // Get all of the cells(2D) or faces(3D) associated with this edge
+    PetscCall(DMPlexGetSupportSize(dm, e, &nFace));
+    PetscCall(DMPlexGetSupport(dm, e, &faces));
+
+    switch (dim) {
+        case 1:
+            N[0] = PetscSignReal(edgeCenter[0] - vCoords[0]);  // In 1D this is simply the vector connecting the vertex to the edge center
+            break;
+        case 2:
+            DMPlexSurfaceAreaNormal2D_Internal(dm, vCoords, edgeCenter, nFace, faces, N);
+            break;
+        case 3:
+            DMPlexSurfaceAreaNormal3D_Internal(dm, -1, vCoords, edgeCenter, nFace, faces, N);
+            break;
+        default:
+            SETERRQ(PetscObjectComm((PetscObject)dm), PETSC_ERR_ARG_OUTOFRANGE, "DMPlexEdgeSurfaceAreaNormal can not handle dimensions of %d", (int)dim);
```

Comment:
You can use the PetscInt_FMT to handle formatting petsc ints when int or long.  This is an example in the code https://github.com/UBCHREST/ablate/blob/350c3a540fd75b7aa33c87e7818a22ce889398f2/src/solver/solver.cpp#L105